### ISSUE TITLE
Enhanced keybord bindings

### ISSUE THREAD

**Issue by reks99** [fesch/Structorizer.Desktop#158]
Hi

as a keyboard junkie I will give an idea how to improve the usage of structorizer.
1. Select a NSD element with cursor keys (up and down for an element and left or right between alternatives as in if, case or parallel).
2. Aktivate the element editor of the selected element with the enter key.
3. Delete the selected element with the delete key.
4. Move the selected element up or down with the page-up or page-down key.
5. Copy and paste a complete nsd from one instance of structorizer into the arranger of another structorizer.

What do you think about this?
Fine regards
Rolf

**Comment by codemanyak** [comment]
w.r.t. 1.: I personally would find it more intuitive to move an element up or down with cursor keys than to shift the selection or otherwise to move buy page navigation keys (as proposed in 4). The sidewards selection jumps might get somewhat vague if the neighbouring branches have different layout i.e. if the element borders are situated at different y coordinates. But it would certainly be implementable (for a moving the effect would be too vaguely predictable).
w.r.t. 2.: Sounds good for me.
w.r.t. 3.: is already implemented. Never tried?
w.r.t. 4.: Isn't very intuitive for me, actually. I would rather reserve these keys for pagewise scrolling.
w.r.t. 5.: Good idea. Should be feasible.

But please don't expect an instant implementation. It's not that urgent in my eyes.

**Comment by codemanyak** [comment]
After some reconsideration I'd follow your proposals 1, 2, and 5 (3 maintained of course), adding Shift-Up and Shift-Down to expand the current selection. and proposing possibly Ctrl-Up/-Down or Alt-Up/-Down, or just "u" and "d" for moving up and down instead of 4. As mentioned earlier, PageUp and PageDown better ought to be used to scroll vertically, I think.

**Comment by reks99** [comment]
@codemanyak Hi, Kay

Your idea using Ctrl-Up/Down sounds even better, than page-up/down.
And I do not expect an instant implementation - it was just a list of ideas which came up.

Fine regards
Rolf

**Comment by codemanyak** [comment]
Okay, I yet already found the some time for a test implementation.
2 and 4 were trivial.
1 turned out to be more tricky until I eventually got hold of the key binding. My approach was to compute the medium x and y coordinates of the currently selected element and then finding the neighbouring element in the respective cursor direction. Works fine, but the behaviour - though fully explainable and correct - might possibly not exactly be what you expect. You must consider the consequences of recursive structural inclusions: whether you may enter a subordinate element of an enclosing structured element depends on the cursor direction! On moving up you would first enter the body of a While or For loop or the branch occupying the central x coordinate of an alternative or Case element and only after having traversed the inner structure the outer one would adopt the selection. Going down, however, you would select the enclosing outer element first and then jump to its successor on the outer level, not enter the substructure.
With Repeat loops it's the other way round: You could only enter the loop body going downwards but not going upwards. The internals of a parallel element are neither reachable going down nor going up because selection would always be caught be the Parallel element as a whole. And so on. Hence, the navigation through a diagram by cursor keys may get somewhat frustrating. So be warned.
5 will have to be postponed.

**Comment by reks99** [comment]
@codemanyak Hi Kay,

IMHO you don't need the x- y- coordiates of the graph of the element. Keyboard cursor movement might be handled as walking through the (linked) list of the NSD-elements (objects). Then a redraw of selected element might be send to the output (draw) function. Mouse cursor needs not to follow the selection.

Ok, all this depends on the implementation.

Perhaps it helps and fine regards
Rolf

**Comment by codemanyak** [comment]
Addendum: Considering the behaviour of the cursor movements according to propsal 1, I withdraw my additional proposal of extending the current solution with Shift+Up or Shift+Down. Since only element sequences or subsequences at a common level may be members of a multiple selection, it would not work in most cases (i.e. if some the consecutive elements are compound ones).

More important: I also thought about a different traversing strategy: Always to dive into the internal structure of selected compound elements when moving downward (IF, CASE, FOR, WHILE, FOREVER??, PARALLEL???) or upwards (REPEAT), respectively. Analogously, on going right from a selected loop or program.
This way, the behaviour feels sound and intuitive now.

**Comment by codemanyak** [comment]
@reks99
> IMHO you don't need the x- y- coordiates of the graph of the element. Keyboard cursor movement might be handled as walking through the (linked) list of the NSD-elements (objects). Then a redraw of selected element might be send to the output (draw) function. Mouse cursor needs not to follow the selection.

Indeed this was my first idea but it soon turned out to induce a 
monstruous algorithm with all the subclass differences and case 
discriminations. The strategic results would have been the same. So I 
opted for the coordinate approach, which consisted of just about a dozen 
lines of code for the first strategy (being slightly less performant, 
which is irrelevant for a key action, though).

> Ok, all this depends on the implementation.

Right. Thanks, nevertheless. Particularly for the great idea about the 
wandering around. It will improve user experience a lot, I'm sure. I'm 
really enthusiastic about the result of the improved traversing strategy.

**Comment by reks99** [comment]
@codemanyak @fesch 
Hello Kay, hello Bob,

I would ask, if there is something new for tests.

Fine regards
Rolf

**Comment by codemanyak** [comment]
I'm just a little short of time at the moment. To initiate a new version only to introduce three of the four proposed new key bindings wouldn't make much sense, would it?

**Comment by reks99** [comment]
@codemanyak Hi Kay,

I can wait.
Rolf

**Comment by codemanyak** [comment]
All proposals implemented now. The copy and paste feature towards an Arranger required some changes in the saving policy (because diagrams copied via the system clipboard don't convey their file context while the transfer is done via the XML representation). Some selection issues in Arranger were fixed on that occasion.
Version 3.24-07 is initiated.

**Comment by reks99** [comment]
@codemanyak Hi Kay.

First of all, many thanks for the new version.
I found a inconsitency when moving an element in an if part. The element can't be moved before the if element.

![test](https://cloud.githubusercontent.com/assets/18035140/14563025/125f55b0-0320-11e6-91a8-e404efeb3382.png)

So in the shown nsd the case element can't be moved by key (ctl-up) before the if element.

Was this intended?
Fine regards
Rolf

**Comment by codemanyak** [comment]
@reks99
Remember that in general an algorithm forms a tree, not a linear list. Indeed, the moving upwards or downwards of an element (or an element subsequence) via keys means to increase or decrease the rank within the instruction sequence it is member of. As it is a linear order change, this operation is restricted to the margins of the respective linear substructure in the algorithm tree.
It has nothing to do with the type of element you move, but all with the context an element is residing in.
In your example, the CASE element can be moved up one step, then it will be at the top of the branch, such that no further raising is possible. This is indeed intended. If you move the instruction `ccc <- "abc"` down one step it will change places with the alternative, which is again intended. If your moving should follow the selection traversal path, it would have to go into the broader branch of the alternative, instead (here it would be the `sonst` branch), which is __*not*__ what was intended by the functionality of moving up and down.

**Comment by codemanyak** [comment]
Just to summarize my essential answer to our question: The actions __move up__ and __move down__ work the way they have always worked and most certainly have been intended. I didn't change the functionality any bit but just added the key binding. I did this before the "wandering selection" was implemented.
I see now that there arose a consistency contradiction: Either the `Ctrl+Up` and `Ctrl+Down` action is consistent to the classic __move-up__ and __move-down__ strategy or it is the the traversing path of the four cursor keys. Both at a time is hardly possible.
I would still rather stick to the currently implemented "traditional" __move-up__ and __move-down__ effect, actually, since it is well-defined and established.
(To move an element to a place far from the sequence it currently belongs to just cut and paste it.)

**Comment by reks99** [comment]
Hi Kay,

Cut'n paste works.

Furhter on I can move an element with the mouse to any possibel positon. So at least the functionallity is available, but not bound to a key.

Fine regards
Rolf

**Comment by codemanyak** [comment]
In addition, you may drag elements with your mouse. This way, you will see some forbidden moves (when source and target turn red). If you could move an element in a key-controlled way along the cursor key traversal path, you would also pass such forbidden zones. If you select e.g. an alternative and move the selection down it would wander into one of the alternative branches; to follow the same path moving the element itself would be illegal - you cannot move an alternative into its own branch. This is to illustrate one of he many calamities you would run into if you try to push an element around beyond its subsequence level by key activities. Another is that the diagram would dramatically change its shape on such moves (and with it the position of the selected and pushed element).
So, if you don't see further deficiencies in the way your proposals were implemented it would be fine if you closed the issue. Thanks in advance.

**Comment by codemanyak** [comment]
Version 3.24-14 will bring two bugfixes / enhancements (#197, #198).

**Comment by codemanyak** [comment]
I will add key bindings for Shift+PageUp (= "page left"), Shift+PageDown (= "page right") as well as Home (Pos1 scroll to top left), End (scroll to bottom right) with version 3.28-13.

**Comment by codemanyak** [comment]
I'm intending to modify the selection expansion slightly in orde to harmonise it with usual GUI behaviour: The initially selected element within a selection span should be remembered as the selection anchor such that vertical cursor movements with Shift key being pressed can also reduce the selected subsequence towards the anchor element. On stepping over it the span would be expanded to other side. Recently, it has only been possible to expand the span, either at top (with Shift-Up) or at bottom (with Shift-Down). This functionality might be preserved but associated to other key bindings, e.g. Alt-Shift-Up and Alt-Shift-Down, respectively. Of course, interferences between the two modes must be considered. The following seems to be a relatively consistent behaviour:
1. The first singular selected element of an element sequence is the "anchor".
1. With a single element being selected, on mouse-clicking with Shift key pressed below it (within the same "subqueue") the selection span will expand downto the now clicked element. The same happens incrementally on pressing Shift-Down. Analogously, the span will expand upwards by clicking on an element above the or incrementally wandering with Shift-Up.
1. With a span being selected, on clicking (with Shift key being pressed) between the anchor element and the lower (or upper) end of the span the span will reduce on this side of the anchor. By clicking on an element outside the span the span will be expanded on that end. The opposite end of the span will not change, even if it reaches beyond the anchor element.
1. With a selection spanning between the anchor element and an element beneath it Shift-Down will expand the selection downwards (as far as possible), Shift-Up will reduce the selection upwards until the anchor element is reached. On passing it the selection will be spanned to the other side (according to 2.). Analogously with a northbound span from the achor.
1. With a span reaching from an element above the anchor to an element below it (i.e. expanded both to North and South) the effect of Shift-Up and Shift-Down will depend upon whether the previous expansion or reduction had been taken place above or below the anchor - it will apply at the same side. If the anchor gets passed by reducing the span via cursor keys then the scan will be reduced further and the next element above or below the previous anchor will be the new anchor - until the border of the span in that direction is reached. Then the anchor will be fixed at that position and rule 2 applies.
1. With Alt-Shift-Up we will always expand at top, with Alt-Shift-Down we will always expand at bottom (as far as possible).
